docs/0003-hashed-n-tuple-trees.md

```diff
@@ -0,0 +1,254 @@
+# OCFL Community Extension 0003: Hashed Truncated N-tuple Trees for OCFL Storage Hierarchies
+
+  * **Extension Name:** 0003-hashed-n-tuple-trees
+  * **Authors:** Peter Winckles
+  * **Minimum OCFL Version:** 1.0
+  * **Obsoletes:** n/a
+  * **Obsoleted by:** n/a
+
+## Overview
+
+This storage root extension describes how to safely map OCFL object identifiers of any length, containing any characters to OCFL object root directories with the primary goals of ensuring portability and filesystem performance at the cost of directory name transparency.
+
+Using this extension, OCFL object identifiers are hashed and encoded as hex strings. These digests are then divided into _N_ n-tuple segments, which are used to create nested paths under the OCFL storage root.
+
+This approach allows OCFL object identifiers of any composition to be evenly distributed across the storage hierarchy. The maximum number of files under any given directory is controlled by the number of characters in each n-tuple, and the tree depth is controlled by the number of n-tuple segments each digest is divided into. Additionally, it obviates the need to handle special characters in OCFL object identifiers because the mapped directory names will only ever contain the characters `0-9a-f`.
+
+However, this comes at the cost of not being able to identify the OCFL object identifier of an object simply by browsing the OCFL storage hierarchy. The ID of an object may only be found within its `inventory.json`.
+
+## Parameters
+
+### Summary
+
+* **Name:** `digestAlgorithm`
+  * **Description:** The digest algorithm to apply on the OCFL object identifier; MUST be an algorithm that is allowed in the OCFL fixity block
+  * **Type:** string
+  * **Range:** 1,4095
```

Comment:
I am not sure I understand the intent of this `1,4095` range definition.

Reply:
I carried that over from @neilsjefferies PR #12. I think the intent is that the value should not be empty. You think smaller arbitrary value would be better?

Reply:
Hmm... I could have sworn that I copied that value from either a max value in the extension spec or another PR, but now I can't find it anywhere. The intent was just that the field should be non-empty. A smaller arbitrary value would also work. Maybe 64?

Reply:
I am surprised that the range is an integer. I would have expected it to be an enum of the allowed digest algorithms.

Reply:
Found the reference! It's in PR #13.

> For string parameters, the range specifies the maximum length of the string as an integer number of characters, not bytes. Again, based on a survey of parsers, strings SHOULD be shorter than 4095 characters

Reply:
Additionally, it can't be a list of algorithms because the allowed algorithms are any algorithm that is allowed to be used in the `fixity` block, which includes the digest algorithm extension.

Reply:
Thanks for digging that up. The `range` of String parameters being a length integer may be something we want to iterate on in the context of: https://github.com/OCFL/extensions/pull/13/files#diff-04c6e90faac2675aa89e2176d2eec7d8R43

..but that is not an issue to be resolved here.